### PR TITLE
[After sales services] refactoring 

### PR DESCRIPTION
**Pull request by atp-odoo**
TASK : https://www.odoo.com/web#id=9372&view_type=form&model=project.task&menu_id=3942&action=327
Pad: https://pad.odoo.com/p/r.jhh4fJAi0f6GvlJ8

### CODE REVIEW

File: addons/crm_claim/crm_claim.py [hunk 1]
```diff
@@ -56,17 +61,68 @@ class crm_claim(osv.osv):
...
+
+        fold = {}
+        for stage in stage_obj.browse(cr, access_rights_uid, stage_ids, context=context):
+            fold[stage.id] = stage.fold or False
+        return result, fold
+
+    def _get_attachment_number(self, cr, uid, ids, fields, args, context=None):
+        res = dict.fromkeys(ids, 0)
+        for app_id in ids:
+            res[app_id] = self.pool['ir.attachment'].search_count(cr, uid, [('res_model', '=', 'crm.claim'), ('res_id', '=', app_id)], context=context)
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: use read group

File: addons/crm_claim/crm_claim.py [hunk 2]
```diff
@@ -56,17 +61,68 @@ class crm_claim(osv.osv):
...
+            'default_partner_id': claim.partner_id and claim.partner_id.id or False,
+            'default_partner_ids': partner_ids,
+        }
+        return res
+
+    def _phonecall_count(self, cr, uid, ids, field_name, arg, context=None):
+        Claim = self.pool['crm.phonecall']
+        return {
+            claim_id: Claim.search_count(cr,uid, [('partner_id', '=', claim_id)], context=context)
+            for claim_id in ids
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: Use Read group instead of search_count

File: addons/crm_claim/crm_claim.py [hunk 3]
```diff
@@ -56,17 +61,68 @@ class crm_claim(osv.osv):
...
+
+    def _get_attachment_number(self, cr, uid, ids, fields, args, context=None):
+        res = dict.fromkeys(ids, 0)
+        for app_id in ids:
+            res[app_id] = self.pool['ir.attachment'].search_count(cr, uid, [('res_model', '=', 'crm.claim'), ('res_id', '=', app_id)], context=context)
+        return res
+
+    def schedule_followup(self, cr, uid, ids, context=None):
+        claim = self.browse(cr, uid, ids[0], context)
+        res = self.pool.get('ir.actions.act_window').for_xml_id(cr, uid, 'calendar', 'action_calendar_event', context)
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: self.pool instead of self.pool.get()

File: addons/crm_claim/crm_claim.py [hunk 4]
```diff
@@ -95,21 +151,46 @@ def _get_default_stage_id(self, cr, uid, context=None):
...
         'partner_phone': fields.char('Phone'),
         'stage_id': fields.many2one ('crm.claim.stage', 'Stage', track_visibility='onchange',
                 domain="['|', ('team_ids', '=', team_id), ('case_default', '=', True)]"),
         'cause': fields.text('Root Cause'),
+        'color': fields.integer('Color Index'),
+        'date_action_last': fields.datetime('Last Action', readonly=1),
+        'date_last_stage_update': fields.datetime('Last Stage Update', select=True),
+        'attachment_number': fields.function(_get_attachment_number, string='Number of Attachments', type="integer"),
+        'phonecall_count': fields.function(_phonecall_count, string="Phonecalls", type="integer"),
+        'followup_ids': fields.one2many('crm.claim.follow', 'claim_id', 'Follow-up', store=True),
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: one2many field with store =True ????

File: addons/crm_claim/crm_claim.py [hunk 5]
```diff
@@ -95,21 +151,46 @@ def _get_default_stage_id(self, cr, uid, context=None):
...
+        'kanban_state': 'normal',
+        'date_last_stage_update': fields.datetime.now,
         'stage_id': lambda s, cr, uid, c: s._get_default_stage_id(cr, uid, c)
     }
+    _group_by_full = {
+        'stage_id': _read_group_stage_ids
+    }
+
+    def action_get_attachment_view(self, cr, uid, ids, context=None):
+        action = self.pool.get('ir.actions.act_window').for_xml_id(cr, uid, 'base', 'action_attachment')
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: use self.pool

File: addons/crm_claim/crm_claim_menu.xml [hunk 6]
```diff
@@ -11,9 +11,10 @@
             <field name="name">Claims</field>
             <field name="res_model">crm.claim</field>
             <field name="view_type">form</field>
-            <field name="view_mode">tree,calendar,form</field>
-            <field name="view_id" ref="crm_case_claims_tree_view"/>
-            <field name="context">{"search_default_user_id":uid, "stage_type":'claim'}</field>
+            <field name="view_mode">kanban,tree,calendar</field>
+            <field name="view_id" eval="False"/>
+            <field name="context">{"search_default_user_id":uid, "stage_type":'claim', 'needaction_menu_ref': 'crm.menu_crm_case_claims'}</field>
+            <field name="domain" eval=""/>
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim_menu.xml]: <field name="domain" eval=""/>???

File: addons/crm_claim/crm_claim.py [hunk 7]
```diff
@@ -202,6 +316,50 @@ def _claim_count(self, cr, uid, ids, field_name, arg, context=None):
...
+        'user_id': fields.many2one('res.users', 'Responsible'),
+        'meeting_id': fields.many2one('calendar.event', 'Meetings'),
+    }
+    _defaults = {
+        'user_id': lambda s, cr, uid, c: uid,
+    }
+
+    def create(self, cr, uid, vals, context=None):
+        context = dict(context or {})
+        meeting_obj = self.pool['calendar.event']
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: give CalendarEvent instead of meering_obj name

File: addons/crm_claim/crm_claim_view.xml [hunk 8]
```diff
@@ -200,6 +219,59 @@
             </field>
         </record>
 
+        <!-- Claim Kanban view -->
+
+        <record model="ir.ui.view" id="crm_case_claims_kanban_view">
+            <field name="name">CRM - Claims Kanban</field>
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim_view.xml]: give crm.claim.kanban name

File: addons/crm_claim/calendar_event.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+import datetime
+
+from openerp import api, fields, models
+
+class CalendarEvent(models.Model):
+    _inherit = "calendar.event"
+
+    claim_id = fields.Many2one('crm.claim', string="Claim", ondelete='cascade')
+    start_datetime = fields.Datetime(default= lambda self: fields.Datetime.now())
```
Comment (hmo-odoo) [inline addons/crm_claim/calendar_event.py]: need lambda function to set default ?

Comment (atp-odoo) [inline addons/crm_claim/calendar_event.py]: yes you are right. lambda not required here.and directly (default= fields.Datetime.now)

**Comment by hmo-odoo** [comment]
continue work on PR: https://github.com/odoo/odoo/pull/7786

**Pull request closed without merging**